persistence/jdbc-common/src/main/java/org/infinispan/persistence/jdbc/common/configuration/ManagedConnectionFactoryConfigurationBuilder.java

```diff
@@ -28,15 +31,22 @@ public AttributeSet attributes() {
       return attributes;
    }
 
-   public void jndiUrl(String jndiUrl) {
+   public ManagedConnectionFactoryConfigurationBuilder<S> jndiUrl(String jndiUrl) {
       attributes.attribute(JNDI_URL).set(jndiUrl);
+      return this;
+   }
+
+   public ManagedConnectionFactoryConfigurationBuilder<S> dataSource(DataSource dataSource) {
+      attributes.attribute(DATA_SOURCE).set(dataSource);
+      return this;
    }
 
    @Override
    public void validate() {
       String jndiUrl = attributes.attribute(JNDI_URL).get();
-      if (jndiUrl == null) {
-         throw new CacheConfigurationException("The jndiUrl has not been specified");
+      DataSource dataSource = attributes.attribute(DATA_SOURCE).get();
+      if (jndiUrl == null && dataSource == null) {
+         throw new CacheConfigurationException("Either jndiUrl or dataSource must be specified");
```

Comment:
Error messages must be implemented in the `Log` interface of the module, so they get assigned an id and may be potentially i18n'ized.
Additionally, it makes no sense to specify both a URL and a DS, so you should throw another exception in that case.